p2p/net/nat/internal/nat/nat.go

```diff
@@ -35,65 +53,91 @@ type NAT interface {
 }
 
-// DiscoverNATs returns all NATs discovered in the network.
-func DiscoverNATs(ctx context.Context) <-chan NAT {
-	nats := make(chan NAT)
+// discoverNATs returns all NATs discovered in the network.
+func discoverNATs(ctx context.Context) ([]NAT, []error) {
+	type natsAndErrs struct {
+		nats []NAT
+		errs []error
+	}
+	upnpCh := make(chan natsAndErrs)
+	pmpCh := make(chan natsAndErrs)
 
 	go func() {
-		defer close(nats)
-
-		upnpIg1 := discoverUPNP_IG1(ctx)
-		upnpIg2 := discoverUPNP_IG2(ctx)
-		natpmp := discoverNATPMP(ctx)
-		upnpGenIGDev := discoverUPNP_GenIGDev(ctx)
-		for upnpIg1 != nil || upnpIg2 != nil || natpmp != nil || upnpGenIGDev != nil {
-			var (
-				nat NAT
-				ok  bool
-			)
-			select {
-			case nat, ok = <-upnpIg1:
-				if !ok {
-					upnpIg1 = nil
-				}
-			case nat, ok = <-upnpIg2:
-				if !ok {
-					upnpIg2 = nil
-				}
-			case nat, ok = <-upnpGenIGDev:
-				if !ok {
-					upnpGenIGDev = nil
-				}
-			case nat, ok = <-natpmp:
-				if !ok {
-					natpmp = nil
-				}
-			case <-ctx.Done():
-				// timeout.
-				return
-			}
-			if ok {
-				select {
-				case nats <- nat:
-				case <-ctx.Done():
-					return
-				}
-			}
+		defer close(upnpCh)
+
+		// We do these UPNP queries sequentially because some routers will fail to handle parallel requests.
+		nats, errs := discoverUPNP_IG1(ctx)
+
+		// Do IG2 after IG1 so that its NAT devices will appear as "better" when we
+		// find the best NAT to return below.
+		n, e := discoverUPNP_IG2(ctx)
+		nats = append(nats, n...)
+		errs = append(errs, e...)
+
+		if len(nats) == 0 {
+			// We don't have a NAT. We should try querying all devices over
+			// SSDP to find a InternetGatewayDevice. This shouldn't be necessary for
+			// a well behaved router.
+			n, e = discoverUPNP_GenIGDev(ctx)
+			nats = append(nats, n...)
+			errs = append(errs, e...)
+		}
+
+		select {
+		case upnpCh <- natsAndErrs{nats, errs}:
+		case <-ctx.Done():
+		}
+	}()
+
+	go func() {
+		defer close(pmpCh)
+		nat, err := discoverNATPMP(ctx)
+		var nats []NAT
+		var errs []error
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			nats = append(nats, nat)
+		}
+		select {
+		case pmpCh <- natsAndErrs{nats, errs}:
+		case <-ctx.Done():
 		}
 	}()
-	return nats
+
+	var nats []NAT
+	var errs []error
+
+	for upnpCh != nil && pmpCh != nil {
+		select {
+		case res := <-pmpCh:
+			pmpCh = nil
+			nats = append(nats, res.nats...)
+			errs = append(errs, res.errs...)
+		case res := <-upnpCh:
+			upnpCh = nil
+			nats = append(nats, res.nats...)
+			errs = append(errs, res.errs...)
+		case <-ctx.Done():
+			errs = append(errs, ctx.Err())
+			return nats, errs
+		}
+	}
+	return nats, errs
 }
 
 // DiscoverGateway attempts to find a gateway device.
 func DiscoverGateway(ctx context.Context) (NAT, error) {
-	var nats []NAT
-	for nat := range DiscoverNATs(ctx) {
-		nats = append(nats, nat)
-	}
+	nats, errs := discoverNATs(ctx)
+
 	switch len(nats) {
 	case 0:
-		return nil, ErrNoNATFound
```

Comment:
We weren't equality checking this error anywhere, were we?

Reply:
no